ttnn/cpp/ttnn/operations/data_movement/repeat_interleave/repeat_interleave.cpp

```diff
@@ -6,51 +6,55 @@
 #include "repeat_interleave.hpp"
 
 #include "ttnn/cpp/ttnn/operations/data_movement/reshape_on_device/reshape.hpp"
+#include "ttnn/cpp/ttnn/operations/data_movement/unsqueeze/unsqueeze.hpp"
+#include "ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp"
+#include "ttnn/cpp/ttnn/operations/copy.hpp"
 
 namespace ttnn {
 namespace operations {
 namespace data_movement {
 
-
 // repeat interleave supports repeats as 1 to inf, dim between 0 to 2
 ttnn::Tensor ExecuteRepeatInterleave::invoke(const ttnn::Tensor& input_a, uint32_t repeat, int32_t dim, std::optional<MemoryConfig> output_mem_config) {
     std::vector<Tensor> combined_tensors;
     combined_tensors.reserve(repeat);
-    auto shape_wh = input_a.get_legacy_shape();
     MemoryConfig mem_config = output_mem_config.value_or(input_a.memory_config());
-    // normalizing the negative dim
-    uint32_t normalized_dim = input_a.get_legacy_shape().get_normalized_index(dim);
-    // check if dim is 1 or 3
-    if (normalized_dim & 1) {
-        constexpr uint32_t tmp_dim = 2;
-        std::vector<int64_t> dims = {0, 1, 2, 3};
-        std::swap(dims[dim], dims[tmp_dim]);
-        Tensor transpose_input = ttnn::permute(input_a, dims);
-        Tensor ril_result = ExecuteRepeatInterleave::invoke(transpose_input, repeat, tmp_dim, mem_config);
-        return ttnn::permute(ril_result, dims);
+    if (repeat == 1) {
+        return ttnn::to_memory_config(input_a, mem_config);
+    }
+    uint32_t input_rank = input_a.get_shape().rank();
+    uint32_t normalized_dim = input_a.get_shape().get_normalized_index(dim);
+    if (normalized_dim == input_rank - 1) {
+        auto transposed_input = ttnn::transpose(input_a, -1, -2, mem_config);
+        auto repeated_input = ExecuteRepeatInterleave::invoke(transposed_input, repeat, -2, mem_config);
+        return ttnn::transpose(repeated_input, -1, -2, mem_config);
+    }
+
+    ttnn::Tensor rm_input = input_a;
+    bool typecast = input_a.get_dtype() != DataType::BFLOAT16;
+    if (typecast) {
+        rm_input = ttnn::typecast(rm_input, DataType::BFLOAT16, mem_config);
```

Comment:
do we do this for unsqueeze?
if so, we should move this to unsqueeze

Reply:
It's not related to unsqueeze. Transpose itself has a limitation on data types for certain transposes besides transpose_wh.

Reply:
I'm a bit confused here. After this optional typecast we do the following:
- To_layout ROW_MAJOR
- Unsqueeze
- Concat
- Reshape
- To_layout original
- Typecast back

Why do we typecast? There's no transpose here. And how is the user of those APIs is supposed to know that they should typecast something?

Reply:
Ok, but the code says `bool typecast = input_a.get_dtype() != DataType::BFLOAT16;`, but you mention the issue is only BFLOAT8_B, so do we actually need the typecast for all non BFLOAT16 types, like FLOAT32 for example?

Reply:
We do for uint16 as well atm due to a lack of support on Grayskulle atm.

Reply:
Basically the thing I'm concerned about is that we're trying to add generality, but `typecast` actually removes generality. Because suddenly ops like `repeat` or `transpose` would lose data precision, which would look insanely strange for the user. But I guess this issue is out of scope of this PR